### PR TITLE
Feature/issue 3344 view panic

### LINKED ISSUE #3344: Sync Gateway 1.5.1 panics when querying a view
This occurs on Sync Gateway 1.5.1. We are currently on 1.4.1, and this functionality works fine. This was found while testing for upgrade path due to other bugs fixed in the 1.5 timeframe.

Occurs on Windows environments. Have tested on Windows 10 as well as Server 2008 R2

Configuration:
Nothing special on the config, using the provided example for pointing at a real bucket, only changing the URL to our server, and the bucket to our bucket. (Also occurs on other environment with multiple buckets currently running 1.4.1 successfully)

Relavent Logs
https://gist.github.com/timritzer/63899082e218029f7d17bc5fe93c5e9b

Expected:
Expected the sync to return filtered view results based on the passed in keys.

What Happens:
Sync Gateway panics logs errors, and fails to respond to request.

Steps to reproduce:

1. Create view through Sync Gateway. We usually do this in the back end application, but for reference created using CURL in windows:
`curl -v -X PUT http://localhost:4985/db/_design/keylookup_withreduce -H "Content-Type: application/json" -d "{ \"views\": {\"keylookup_withreduce\": { \"map\": \"function (doc, meta) { if(doc._deleted) { return; } for(var i in doc.channels) { emit(doc.channels[i], doc.keyid);}}\" , \"reduce\": \"function (keys, values, rereduce) { if (rereduce) { var result = []; for (var i = 0; i ^< values.length; i++) { var distinct = values[i]; for (var j = 0; j ^< distinct.length; j++){result.push(distinct[j]);}} return result.filter(function (e, i, arr) { return arr.lastIndexOf(e) === i;}); } return values.filter(function (e, i, arr) { return arr.lastIndexOf(e) === i; } );} \" }}}"`

2. Query view through Sync Gateway. Again, usually done in code, but for purposes of recreation here is a CURL request:
`curl -v http://localhost:4985/db/_design/keylookup_withreduce/_view/keylookup_withreduce?keys=%5B%22env_1%22%5D`

3. Note that no response is returned, and the error logs show the panic posted in the logs above.

Note that querying for only one key works as expected, and returns results.
`curl -v http://localhost:4985/db/_design/keylookup_withreduce/_view/keylookup_withreduce?key=%22env_1%22`
Also querying through Couchbase itself works as well.

### PR DESCRIPTION
**Pull request by tleyden**
Fixes #3344 

Integration tests:

http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/451/
http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/452/

### CODE REVIEW

**Comment by tleyden** [comment]
I'm seeing test failures:

```
--- FAIL: TestApplyViewQueryOptions (0.00s)
panic: interface conversion: interface {} is []string, not []interface {} [recovered]
	panic: interface conversion: interface {} is []string, not []interface {}
```

Looking into this.

**Comment by tleyden** [comment]
I updated the PR to handle both `[]interface{}` and `[]string`.

Kicked off integration tests again

**Comment by tleyden** [comment]
1. http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/452/ passed
1. http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/451/ failed, but due to [known sporadic failure](https://github.com/couchbase/sync_gateway/issues/3006) 

Pushed a new commit w/ some refactoring to improve the codebeat score

**Comment by tleyden** [comment]
@adamcfraser this one is ready for review.  The drone CI build passed, but somehow the github issue didn't get updated (click details)

File: base/util.go [hunk 1]
```diff
@@ -826,3 +826,21 @@ func ConvertBackQuotedStrings(data []byte) []byte {
...
+	switch v := i.(type) {
+	case []string:
+		emptyInterfaceKeys := []interface{}{}
+		for _, key := range v {
+			emptyInterfaceKeys = append(emptyInterfaceKeys, key)
+		}
+		return emptyInterfaceKeys, nil
+	case []interface{}:
+		return v, nil
+	default:
```
Comment (adamcfraser) [inline base/util.go]: Are we sure that []string and []interface{} are the only cases we need to support?  No []int, non-slice, etc?

**Review by adamcfraser — COMMENTED** [review]
Question on whether there are other possibilities we need to support - other than that it looks good.

File: base/util.go [hunk 2]
```diff
@@ -826,3 +826,21 @@ func ConvertBackQuotedStrings(data []byte) []byte {
 		return bytes
 	})
 }
+
+
+func ConvertToEmptyInterfaceSlice(i interface{}) (result []interface{}, err error) {
+
+	switch v := i.(type) {
+	case []string:
+		emptyInterfaceKeys := []interface{}{}
```
Comment (bbrks) [inline base/util.go]: We could pre-allocate the length for the slice in here:

```go
var emptyInterfaceKeys = make([]interface{}, len(v))
for i, data range v {
	emptyInterfaceKeys[i] = data
}
return emptyInterfaceKeys, nil
```

**Comment by adamcfraser** [comment]
Made changes based on PR feedback, queued up new integration tests:
http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/453/
http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/454/

**Comment by adamcfraser** [comment]
Need to review handling for composite or otherwise non-string keys (e.g. use channel view case as example)

**Comment by adamcfraser** [comment]
Added additional tests to validate handling for non-string keys.

http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/455/
http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/456/

**Comment by adamcfraser** [comment]
Integration tests are failing intermittently on new tests with error like:
500 Internal error: unknown_error - view_undefined

This is down to the design doc being ready but the view not being ready when the query is issued.  Need to add a test helper that retries on view_undefined until the view is available.  The existing 'WaitForNAdminViewResults' will not work - it errors out when getting a 500.

**Comment by adamcfraser** [comment]
Added new Wait utility method, reran integration tests, passing now:
http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/457/
http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/458/

**Review by adamcfraser — APPROVED** [review]

**Pull request merged**